Ekr editorial auth48 (#134)
quicwg/version-negotiation: draft-ietf-quic-version-negotiation.md

server MAY add reserved versions (as defined in {{Section 6.3 of QUIC}}) in
Supported Version fields.

> Clients MUST ignore a Version Negotiation packet if it contains the Original

Author: I'm not comfortable with a naked "will" like this. If you want to cite to the downgrade section, I'm fine with that.

> Clients MUST ignore a Version Negotiation packet if it contains the Original

Reviewer: Don't we cite that section in that sentence? Keeping `will` and replacing `(see {{downgrade}})` with `, as required by {{downgrade}}` SGTM though

> Version attempted by the client (see {{downgrade}}). The client MUST also ignore a

Reviewer: Similarly, this was intentional to avoid repeating 8999

Author: See above. My problem is that naked statements of how the client behaves need to be supported by something. If you want to put (as required by...) that would be fine.

version negotiation errors are signaled using a transport error of type
VERSION_NEGOTIATION_ERROR (see {{iana-error}}).

When a server receives a client's first flight, the server will first
establish which QUIC version is in use for this connection in order to

> properly parse the first flight. This may involve examining packets
> that are not part of the handshake transcript, such as in the the

Reviewer: ```suggestion
properly parse the first flight. This may involve examining data
that is not part of the handshake transcript, such as parts of the
```

When a server receives a client's first flight, the server will first establish
which QUIC version is in use for this connection in order to properly parse the

> first flight. This may involve examining packets outside of the handshake, such

Reviewer: I'm not sure what "examining packets outside of the handshake" means. INITIAL packets are still part of the handshake in the generic sense.

I suspect you meant something like "examining data that is not covered by the handshake transcript" but that's a mouthful and also quite specific to v1

How about just "This may involve examining the long header Version field."?

Author: Well, it might involve examining something else. E.g., the length of the packet.

Reviewer: Agreed. But I still don't understand what "packets outside of the handshake" means

Author: PTAL

When a server receives a client's first flight, the server will first establish
which QUIC version is in use for this connection in order to properly parse the
first flight. This may involve examining packets outside of the handshake, such
as in the the packet header. When the server then processes the client's Version
Information, the server MUST validate that the client's Chosen Version matches
the version in use for the connection. If the two differ, the server MUST close
the connection with a version negotiation error.

> In the specific case of QUIC version 1, the client determines that it is in use

Reviewer: ```suggestion
In the specific case of QUIC version 1, the server determines that version 1 is in use
```

by observing that the Version field of the first Long Header packet it receives

> is set to 0x00000001. Subsequently, if the client receives the client's Version

Reviewer: ```suggestion
is set to 0x00000001. Subsequently, if the server receives the client's Version
```

Author: Hmm.. This paragraph or not, it seems that *both* endpoints look at the long header, no?

Reviewer: Well yeah, but the server is the one reading the client's Version

Information over QUIC version 1 (as indicated by the Version field of the Long
Header packets that carried the transport parameters) and the client's Chosen
Version is not set to 0x00000001, the server MUST close the connection with a
version negotiation error.

> Servers MAY complete the handshake even if the Version Information is missing.

Reviewer: For this paragraph I personally find the previous phrasing easier to reason about, but I don't feel too strongly

Author: I found the ordering pretty confusing. Happy to look at yet another alternative if you want.

Reviewer: This ordering works for me, let's keep yours.

Clients MUST NOT complete the handshake if they are reacting to a Version
Negotiation packet and the Version Information is missing, but MAY do so
otherwise.

> If a client receives Version Information where the server's Chosen Version was

Reviewer: Good catch!

not sent by the client as part of its Available Versions, the client MUST close
the connection with a version negotiation error. If a client has reacted to a

> Version Negotiation packet and the Version Information was missing, the client

Reviewer: ```suggestion
Version Negotiation packet and the server's Version Information was missing, the client
```

MUST close the connection with a version negotiation error.

> 

Reviewer: GitHub is silly so I can't comment further down, but can you s/is made aware/can be made aware/ on line 479?

Similarly, s/will/MUST/ on line 657
